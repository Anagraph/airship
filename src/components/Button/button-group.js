import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { darken } from 'polished';
import Button from './button';
import { isComponentOfType } from '../../utils';
import { colors } from '../../constants';

const separator = (props) => {
  return `1px solid ${
    !!props.secondary ? colors.primaryColor : darken(0.16, colors.primaryColor)
  }`;
};

const StyledList = styled.ul`
  border: 1px solid ${colors.primaryColor};
  border-radius: 4px;
  display: inline-flex;
  list-style: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
`;

const StyledListItem = styled.li`
  border-right: ${separator};
  display: flex;

  &:last-child {
    border-right: 0;
  }
`;

class ButtonGroup extends Component {
  render() {
    const { children, secondary, ...other } = this.props;
    const buttons = React.Children.map(children, (child) => {
      return (
        <StyledListItem secondary={secondary}>
          {isComponentOfType(Button, child)
            ? React.cloneElement(child, {
                radius: 0,
                grouped: true,
                borderless: !!secondary,
                secondary: false
              })
            : child}
        </StyledListItem>
      );
    });

    return (
      <StyledList {...other} data-component="ButtonGroup">
        {buttons}
      </StyledList>
    );
  }
}

ButtonGroup.propTypes = {
  secondary: PropTypes.bool
};

export default ButtonGroup;
