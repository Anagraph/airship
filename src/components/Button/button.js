import React, { Component } from 'react';
import styled from 'styled-components';
import { darken, lighten } from 'polished';
import { colors } from '../../constants';

const font = (props) => {
  let font = "500 12px/20px 'Roboto'";
  if (props.large) {
    font = "500 16px/24px 'Roboto'";
  }
  if (props.small) {
    font = "400 10px/12px 'Roboto'";
  }
  return font;
};

const padding = (props) => {
  let padding = '6px 12px';
  if (props.large) {
    padding = '8px 16px';
  }
  if (props.small) {
    padding = '6px 8px';
  }
  return padding;
};

const background = (props) => {
  return !!props.borderless || !!props.secondary
    ? 'transparent'
    : colors.primaryColor;
};

const radius = (props) => {
  return !!props.borderless ? 0 : '4px';
};

const color = (props) => {
  return !!props.borderless || !!props.secondary
    ? colors.primaryColor
    : colors.white;
};

const border = (props) => {
  return !!props.secondary ? `1px solid ${colors.primaryColor}` : 0;
};

const hover = (props) => {
  return !!props.borderless || !!props.secondary
    ? lighten(0.45, colors.primaryColor)
    : darken(0.16, colors.primaryColor);
};

const focus = (props) => {
  return !!props.borderless || !!props.secondary
    ? lighten(0.4, colors.primaryColor)
    : darken(0.24, colors.primaryColor);
};

const size = (props) => {
  return !!props.large ? 16 : 12;
};

const StyledButton = styled.button`
  border: ${border};
  box-shadow: none;
  border-radius: ${radius};
  background: ${background};
  color: ${color};
  cursor: pointer;
  padding: ${padding};
  font: ${font};
  display: flex;
  align-items: center;

  &[disabled],
  &.is-disabled {
    opacity: 0.24;
    cursor: default;
  }

  &:hover {
    background: ${hover};
  }

  &:focus {
    background: ${focus};
  }

  &.is-disabled:hover,
  &.is-disabled:focus {
    background: ${background};
    color: ${color};
  }

  svg {
    &:only-child {
      margin: 0 -6px;
    }
  }

  svg + * {
    margin-left: 8px;
  }
`;

class Button extends Component {
  renderChildren() {
    const { props } = this;
    return React.Children.map(props.children, (child) => {
      return typeof child === 'string' ? (
        <span>{child}</span>
      ) : (
        // what if it's not svg?
        React.cloneElement(child, {
          className: 'button-media',
          color: color(props),
          width: size(props),
          height: size(props)
        })
      );
    });
  }

  render() {
    return (
      <StyledButton {...this.props} data-component="Button">
        {this.renderChildren()}
      </StyledButton>
    );
  }
}
export default Button;
