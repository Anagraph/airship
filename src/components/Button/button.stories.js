import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { ThemeProvider } from 'styled-components';
import Button from './button';
import ButtonGroup from './button-group';
import Icon from '../Icon/icon';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#FABADA',
};

storiesOf('Button', module)
  .add('Default', () => (
    <div>
      <h3 className="header">Primary</h3>
      <Button>Click me</Button>

      <h3 className="header">Scondary</h3>
      <Button secondary>Click me</Button>

      <h3 className="header">Borderless</h3>
      <Button borderless>Click me</Button>

      <h3 className="header">With event</h3>
      <Button onClick={action('click')}>Click me</Button>

      <h3 className="header">Grouped</h3>
      <ButtonGroup>
        <Button>
          <Icon icon="plus" />
        </Button>
        <Button>Edit</Button>
        <Button>
          Delete
        </Button>
      </ButtonGroup>

      <h3 className="header">Grouped Secondary</h3>
      <ButtonGroup secondary>
        <Button>
          <Icon icon="plus" />
        </Button>
        <Button>Edit</Button>
        <Button>
          Delete
        </Button>
      </ButtonGroup>
    </div>
  ))
  .add('Disabled', () => (
    <div>
      <h3 className="header">Primary</h3>
      <Button disabled>Click me</Button>

      <h3 className="header">Scondary</h3>
      <Button disabled secondary>Click me</Button>

      <h3 className="header">Borderless</h3>
      <Button disabled borderless>Click me</Button>

      <h3 className="header">With event</h3>
      <Button disabled onClick={action('click')}>Click me</Button>

      <h3 className="header">Grouped</h3>
      <ButtonGroup>
        <Button disabled>
          <Icon icon="plus" />
        </Button>
        <Button disabled>Edit</Button>
        <Button disabled>
          Delete
        </Button>
      </ButtonGroup>

      <h3 className="header">Grouped Secondary</h3>
      <ButtonGroup secondary>
        <Button disabled>
          <Icon icon="plus" />
        </Button>
        <Button disabled>Edit</Button>
        <Button disabled>
          Delete
        </Button>
      </ButtonGroup>
    </div>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <div>
        <h3 className="header">Primary</h3>
        <Button>Click me</Button>

        <h3 className="header">Scondary</h3>
        <Button secondary>Click me</Button>

        <h3 className="header">Borderless</h3>
        <Button borderless>Click me</Button>

        <h3 className="header">With event</h3>
        <Button onClick={action('click')}>Click me</Button>

        <h3 className="header">Grouped</h3>
        <ButtonGroup>
          <Button>
            <Icon icon="plus" />
          </Button>
          <Button>Edit</Button>
          <Button>
            Delete
          </Button>
        </ButtonGroup>

        <h3 className="header">Grouped Secondary</h3>
        <ButtonGroup secondary>
          <Button>
            <Icon icon="plus" />
          </Button>
          <Button>Edit</Button>
          <Button>
            Delete
          </Button>
        </ButtonGroup>
      </div>
    </ThemeProvider>
  ));
