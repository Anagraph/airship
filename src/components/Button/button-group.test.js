import React from 'react';
import renderer from 'react-test-renderer';
import ButtonGroup from './button-group';
import Button from './button';
import PlusIcon from '../Icons/plus';

describe('render', () => {
  it('renders without crashing', () => {
    const component = renderer.create(
      <ButtonGroup>
        <Button>
          <PlusIcon />
        </Button>
        <Button>Mundo</Button>
        <Button>Happy</Button>
      </ButtonGroup>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders secondary button', () => {
    const component = renderer.create(
      <ButtonGroup secondary>
        <Button>
          <PlusIcon />
        </Button>
        <Button>Mundo</Button>
        <Button>Happy</Button>
      </ButtonGroup>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
