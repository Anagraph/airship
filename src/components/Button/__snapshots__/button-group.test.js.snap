// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders secondary button 1`] = `
.c0 {
  border: 1px solid #1785FB;
  border-radius: 4px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  list-style: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.c1 {
  border-right: 1px solid #1785FB;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1:last-child {
  border-right: 0;
}

.c2 {
  border: 0;
  box-shadow: none;
  border-radius: 0px;
  background: transparent;
  color: #1785FB;
  cursor: pointer;
  padding: 6px 12px;
  font: 500 12px/20px 'Roboto';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2[disabled],
.c2.is-disabled {
  opacity: 0.24;
  cursor: default;
}

.c2:hover {
  background: #f9fcff;
}

.c2:focus {
  background: #e0eefe;
  outline: none;
}

.c2.is-disabled:hover,
.c2.is-disabled:focus {
  background: transparent;
  color: #1785FB;
}

.c2 .button-media {
  margin: 0 -6px;
}

.c2 svg + * {
  margin-left: 8px;
}

<ul
  className="c0"
  data-component="ButtonGroup"
>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <svg
        className=""
        fill="#1785FB"
        height={16}
        viewBox="0 0 15 15"
        width={16}
      >
        <polygon
          points="0.020904541 7.00476178 7.01217651 7.00476178 7.06573917 0.0133888789 8.00093254 0.0133888789 8.00093254 7.00476178 15.0067139 7.00476178 15.0067139 8.00007932 8.00093254 8.00007932 8.00093254 14.9849872 7.01217651 14.9849872 7.01217651 8.00007932 0.020904541 8.00007932"
        />
      </svg>
    </button>
  </li>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <span>
        Mundo
      </span>
    </button>
  </li>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <span>
        Happy
      </span>
    </button>
  </li>
</ul>
`;

exports[`render renders without crashing 1`] = `
.c0 {
  border: 1px solid #1785FB;
  border-radius: 4px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  list-style: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
}

.c1 {
  border-right: 1px solid #035dbd;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1:last-child {
  border-right: 0;
}

.c2 {
  border: 0;
  box-shadow: none;
  border-radius: 0px;
  background: #1785FB;
  color: #fff;
  cursor: pointer;
  padding: 6px 12px;
  font: 500 12px/20px 'Roboto';
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2[disabled],
.c2.is-disabled {
  opacity: 0.24;
  cursor: default;
}

.c2:hover {
  background: #035dbd;
}

.c2:focus {
  background: #034995;
  outline: none;
}

.c2.is-disabled:hover,
.c2.is-disabled:focus {
  background: #1785FB;
  color: #fff;
}

.c2 .button-media {
  margin: 0 -6px;
}

.c2 svg + * {
  margin-left: 8px;
}

<ul
  className="c0"
  data-component="ButtonGroup"
>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <svg
        className=""
        fill="#fff"
        height={16}
        viewBox="0 0 15 15"
        width={16}
      >
        <polygon
          points="0.020904541 7.00476178 7.01217651 7.00476178 7.06573917 0.0133888789 8.00093254 0.0133888789 8.00093254 7.00476178 15.0067139 7.00476178 15.0067139 8.00007932 8.00093254 8.00007932 8.00093254 14.9849872 7.01217651 14.9849872 7.01217651 8.00007932 0.020904541 8.00007932"
        />
      </svg>
    </button>
  </li>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <span>
        Mundo
      </span>
    </button>
  </li>
  <li
    className="c1"
  >
    <button
      className="c2"
      data-component="Button"
      radius={0}
    >
      <span>
        Happy
      </span>
    </button>
  </li>
</ul>
`;
